#NLP AND NAIVE BAYES

In [1]:
import nltk
nltk.download('all')

[nltk_data] Downloading collection 'all'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/abc.zip.
[nltk_data]    | Downloading package alpino to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/alpino.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_eng to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping
[nltk_data]    |       taggers/averaged_perceptron_tagger_eng.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_ru to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping
[nltk_data]    |       taggers/averaged_perceptron_tagger_ru.zip.
[nltk_data]    | Downloading package averaged_perceptron_tagger_rus to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |  

True

#Data Exploration and Preprocessing

In [2]:
import pandas as pd
import numpy as np
df=pd.read_csv("blogs.csv")
df

Data              Labels
0     Path: cantaloupe.srv.cs.cmu.edu!magnesium.club...         alt.atheism
1     Newsgroups: alt.atheism\nPath: cantaloupe.srv....         alt.atheism
2     Path: cantaloupe.srv.cs.cmu.edu!das-news.harva...         alt.atheism
3     Path: cantaloupe.srv.cs.cmu.edu!magnesium.club...         alt.atheism
4     Xref: cantaloupe.srv.cs.cmu.edu alt.atheism:53...         alt.atheism
...                                                 ...                 ...
1995  Xref: cantaloupe.srv.cs.cmu.edu talk.abortion:...  talk.religion.misc
1996  Xref: cantaloupe.srv.cs.cmu.edu talk.religion....  talk.religion.misc
1997  Xref: cantaloupe.srv.cs.cmu.edu talk.origins:4...  talk.religion.misc
1998  Xref: cantaloupe.srv.cs.cmu.edu talk.religion....  talk.religion.misc
1999  Xref: cantaloupe.srv.cs.cmu.edu sci.skeptic:43...  talk.religion.misc

[2000 rows x 2 columns]

In [3]:
#tokenization
from nltk.tokenize import word_tokenize

df['Tokenized_Data'] = df['Data'].apply(word_tokenize)
display(df.head())

Data       Labels  \
0  Path: cantaloupe.srv.cs.cmu.edu!magnesium.club...  alt.atheism   
1  Newsgroups: alt.atheism\nPath: cantaloupe.srv....  alt.atheism   
2  Path: cantaloupe.srv.cs.cmu.edu!das-news.harva...  alt.atheism   
3  Path: cantaloupe.srv.cs.cmu.edu!magnesium.club...  alt.atheism   
4  Xref: cantaloupe.srv.cs.cmu.edu alt.atheism:53...  alt.atheism   

                                      Tokenized_Data  
0  [Path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...  
1  [Newsgroups, :, alt.atheism, Path, :, cantalou...  
2  [Path, :, cantaloupe.srv.cs.cmu.edu, !, das-ne...  
3  [Path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...  
4  [Xref, :, cantaloupe.srv.cs.cmu.edu, alt.athei...

In [4]:
#normalization:convert all words to lowercase
df['Normalized_Data'] = df['Tokenized_Data'].apply(lambda x: [word.lower() for word in x])
display(df.head())

Data       Labels  \
0  Path: cantaloupe.srv.cs.cmu.edu!magnesium.club...  alt.atheism   
1  Newsgroups: alt.atheism\nPath: cantaloupe.srv....  alt.atheism   
2  Path: cantaloupe.srv.cs.cmu.edu!das-news.harva...  alt.atheism   
3  Path: cantaloupe.srv.cs.cmu.edu!magnesium.club...  alt.atheism   
4  Xref: cantaloupe.srv.cs.cmu.edu alt.atheism:53...  alt.atheism   

                                      Tokenized_Data  \
0  [Path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
1  [Newsgroups, :, alt.atheism, Path, :, cantalou...   
2  [Path, :, cantaloupe.srv.cs.cmu.edu, !, das-ne...   
3  [Path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
4  [Xref, :, cantaloupe.srv.cs.cmu.edu, alt.athei...   

                                     Normalized_Data  
0  [path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...  
1  [newsgroups, :, alt.atheism, path, :, cantalou...  
2  [path, :, cantaloupe.srv.cs.cmu.edu, !, das-ne...  
3  [path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...  
4  [xref, :, cantaloupe.srv.cs.cmu.edu, alt.athei...

In [5]:
#removing punctuation
import string
df['Punctuation_Removed'] = df['Normalized_Data'].apply(lambda x: [word for word in x if word not in string.punctuation])
display(df.head())

Data       Labels  \
0  Path: cantaloupe.srv.cs.cmu.edu!magnesium.club...  alt.atheism   
1  Newsgroups: alt.atheism\nPath: cantaloupe.srv....  alt.atheism   
2  Path: cantaloupe.srv.cs.cmu.edu!das-news.harva...  alt.atheism   
3  Path: cantaloupe.srv.cs.cmu.edu!magnesium.club...  alt.atheism   
4  Xref: cantaloupe.srv.cs.cmu.edu alt.atheism:53...  alt.atheism   

                                      Tokenized_Data  \
0  [Path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
1  [Newsgroups, :, alt.atheism, Path, :, cantalou...   
2  [Path, :, cantaloupe.srv.cs.cmu.edu, !, das-ne...   
3  [Path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
4  [Xref, :, cantaloupe.srv.cs.cmu.edu, alt.athei...   

                                     Normalized_Data  \
0  [path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
1  [newsgroups, :, alt.atheism, path, :, cantalou...   
2  [path, :, cantaloupe.srv.cs.cmu.edu, !, das-ne...   
3  [path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
4  [xref, :, cantaloupe.srv.cs.cmu.edu, alt.athei...   

                                 Punctuation_Removed  
0  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...  
1  [newsgroups, alt.atheism, path, cantaloupe.srv...  
2  [path, cantaloupe.srv.cs.cmu.edu, das-news.har...  
3  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...  
4  [xref, cantaloupe.srv.cs.cmu.edu, alt.atheism:...

In [6]:
#removing stop words
from nltk.corpus import stopwords
stop_words = set(stopwords.words('english'))
df['Stop_Words_Removed'] = df['Punctuation_Removed'].apply(lambda x: [word for word in x if word not in stop_words])
display(df.head())

Data       Labels  \
0  Path: cantaloupe.srv.cs.cmu.edu!magnesium.club...  alt.atheism   
1  Newsgroups: alt.atheism\nPath: cantaloupe.srv....  alt.atheism   
2  Path: cantaloupe.srv.cs.cmu.edu!das-news.harva...  alt.atheism   
3  Path: cantaloupe.srv.cs.cmu.edu!magnesium.club...  alt.atheism   
4  Xref: cantaloupe.srv.cs.cmu.edu alt.atheism:53...  alt.atheism   

                                      Tokenized_Data  \
0  [Path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
1  [Newsgroups, :, alt.atheism, Path, :, cantalou...   
2  [Path, :, cantaloupe.srv.cs.cmu.edu, !, das-ne...   
3  [Path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
4  [Xref, :, cantaloupe.srv.cs.cmu.edu, alt.athei...   

                                     Normalized_Data  \
0  [path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
1  [newsgroups, :, alt.atheism, path, :, cantalou...   
2  [path, :, cantaloupe.srv.cs.cmu.edu, !, das-ne...   
3  [path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
4  [xref, :, cantaloupe.srv.cs.cmu.edu, alt.athei...   

                                 Punctuation_Removed  \
0  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...   
1  [newsgroups, alt.atheism, path, cantaloupe.srv...   
2  [path, cantaloupe.srv.cs.cmu.edu, das-news.har...   
3  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...   
4  [xref, cantaloupe.srv.cs.cmu.edu, alt.atheism:...   

                                  Stop_Words_Removed  
0  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...  
1  [newsgroups, alt.atheism, path, cantaloupe.srv...  
2  [path, cantaloupe.srv.cs.cmu.edu, das-news.har...  
3  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...  
4  [xref, cantaloupe.srv.cs.cmu.edu, alt.atheism:...

In [7]:
#lemmatization
from nltk.stem import WordNetLemmatizer
lemmatizer = WordNetLemmatizer()
df['Lemmatized_Data'] = df['Stop_Words_Removed'].apply(lambda x: [lemmatizer.lemmatize(word) for word in x])
display(df.head())

Data       Labels  \
0  Path: cantaloupe.srv.cs.cmu.edu!magnesium.club...  alt.atheism   
1  Newsgroups: alt.atheism\nPath: cantaloupe.srv....  alt.atheism   
2  Path: cantaloupe.srv.cs.cmu.edu!das-news.harva...  alt.atheism   
3  Path: cantaloupe.srv.cs.cmu.edu!magnesium.club...  alt.atheism   
4  Xref: cantaloupe.srv.cs.cmu.edu alt.atheism:53...  alt.atheism   

                                      Tokenized_Data  \
0  [Path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
1  [Newsgroups, :, alt.atheism, Path, :, cantalou...   
2  [Path, :, cantaloupe.srv.cs.cmu.edu, !, das-ne...   
3  [Path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
4  [Xref, :, cantaloupe.srv.cs.cmu.edu, alt.athei...   

                                     Normalized_Data  \
0  [path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
1  [newsgroups, :, alt.atheism, path, :, cantalou...   
2  [path, :, cantaloupe.srv.cs.cmu.edu, !, das-ne...   
3  [path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
4  [xref, :, cantaloupe.srv.cs.cmu.edu, alt.athei...   

                                 Punctuation_Removed  \
0  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...   
1  [newsgroups, alt.atheism, path, cantaloupe.srv...   
2  [path, cantaloupe.srv.cs.cmu.edu, das-news.har...   
3  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...   
4  [xref, cantaloupe.srv.cs.cmu.edu, alt.atheism:...   

                                  Stop_Words_Removed  \
0  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...   
1  [newsgroups, alt.atheism, path, cantaloupe.srv...   
2  [path, cantaloupe.srv.cs.cmu.edu, das-news.har...   
3  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...   
4  [xref, cantaloupe.srv.cs.cmu.edu, alt.atheism:...   

                                     Lemmatized_Data  
0  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...  
1  [newsgroups, alt.atheism, path, cantaloupe.srv...  
2  [path, cantaloupe.srv.cs.cmu.edu, das-news.har...  
3  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...  
4  [xref, cantaloupe.srv.cs.cmu.edu, alt.atheism:...

In [8]:
#feature extraction
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer()
corpus = [' '.join(words) for words in df['Lemmatized_Data']]
X = vectorizer.fit_transform(corpus)
feature_names = vectorizer.get_feature_names_out()
df_features = pd.DataFrame(X.toarray(), columns=feature_names)
display(df_features.head())

00  000  0000  00000  0000001200  000003  000005102000  00000510200001  \
0   0    0     0      0           0       0             0               0   
1   0    0     0      0           0       0             0               0   
2   0    0     0      0           0       0             0               0   
3   0    1     0      0           0       0             0               0   
4   0    0     0      0           0       0             0               0   

   00000f  000036  ...  zy  zy4uj  zyda  zyeh  zyxel  zz  zz0  zz2q1  zzs  zzx  
0       0       0  ...   0      0     0     0      0   0    0      0    0    0  
1       0       0  ...   0      0     0     0      0   0    0      0    0    0  
2       0       0  ...   0      0     0     0      0   0    0      0    0    0  
3       0       0  ...   0      0     0     0      0   0    0      0    0    0  
4       0       0  ...   0      0     0     0      0   0    0      0    0    0  

[5 rows x 48693 columns]

#Naive Bayes Model for Text Classification

In [9]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(df_features, df['Labels'], test_size=0.2, random_state=42)
print("Training set shape:", X_train.shape)
print("Testing set shape:", X_test.shape)
print("Training labels shape:", y_train.shape)
print("Testing labels shape:", y_test.shape)

Training set shape: (1600, 48693)
Testing set shape: (400, 48693)
Training labels shape: (1600,)
Testing labels shape: (400,)


In [10]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

clf = MultinomialNB()
clf.fit(X_train, y_train)

MultinomialNB()

In [11]:
y_pred = clf.predict(X_test)

accuracy_score=accuracy_score(y_test,y_pred)
print("Accuracy Score:",accuracy_score)

Accuracy Score: 0.79


#Sentiment Analysis

In [12]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()
df['Sentiment_Score'] = df['Data'].apply(lambda x: analyzer.polarity_scores(x)['compound'])
df['Sentiment_Label'] = df['Sentiment_Score'].apply(lambda x: 'Positive' if x > 0 else ('Negative' if x < 0 else 'Neutral'))
display(df.head())

Data       Labels  \
0  Path: cantaloupe.srv.cs.cmu.edu!magnesium.club...  alt.atheism   
1  Newsgroups: alt.atheism\nPath: cantaloupe.srv....  alt.atheism   
2  Path: cantaloupe.srv.cs.cmu.edu!das-news.harva...  alt.atheism   
3  Path: cantaloupe.srv.cs.cmu.edu!magnesium.club...  alt.atheism   
4  Xref: cantaloupe.srv.cs.cmu.edu alt.atheism:53...  alt.atheism   

                                      Tokenized_Data  \
0  [Path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
1  [Newsgroups, :, alt.atheism, Path, :, cantalou...   
2  [Path, :, cantaloupe.srv.cs.cmu.edu, !, das-ne...   
3  [Path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
4  [Xref, :, cantaloupe.srv.cs.cmu.edu, alt.athei...   

                                     Normalized_Data  \
0  [path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
1  [newsgroups, :, alt.atheism, path, :, cantalou...   
2  [path, :, cantaloupe.srv.cs.cmu.edu, !, das-ne...   
3  [path, :, cantaloupe.srv.cs.cmu.edu, !, magnes...   
4  [xref, :, cantaloupe.srv.cs.cmu.edu, alt.athei...   

                                 Punctuation_Removed  \
0  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...   
1  [newsgroups, alt.atheism, path, cantaloupe.srv...   
2  [path, cantaloupe.srv.cs.cmu.edu, das-news.har...   
3  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...   
4  [xref, cantaloupe.srv.cs.cmu.edu, alt.atheism:...   

                                  Stop_Words_Removed  \
0  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...   
1  [newsgroups, alt.atheism, path, cantaloupe.srv...   
2  [path, cantaloupe.srv.cs.cmu.edu, das-news.har...   
3  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...   
4  [xref, cantaloupe.srv.cs.cmu.edu, alt.atheism:...   

                                     Lemmatized_Data  Sentiment_Score  \
0  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...          -0.9966   
1  [newsgroups, alt.atheism, path, cantaloupe.srv...           0.8856   
2  [path, cantaloupe.srv.cs.cmu.edu, das-news.har...          -0.9969   
3  [path, cantaloupe.srv.cs.cmu.edu, magnesium.cl...          -0.9999   
4  [xref, cantaloupe.srv.cs.cmu.edu, alt.atheism:...           0.9821   

  Sentiment_Label  
0        Negative  
1        Positive  
2        Negative  
3        Negative  
4        Positive

#Evaluation

* The Naive Bayes model achived 0.79 accuracy which means 79% data is accurately classified which is a good model and there are no challenges that I've came across,although one potential challenge is the high dimensionality because of more number of features.


* Using the sentiment analyser I calculated the sentiment analysis for each blog posts.I used polarity scores which is between -1 to +1 to and calculated the score for every individual blog post and then classified them as positive and negative based on the scores achieved by giving a cutoff score, if it is less than zero then it is negative and if it is greater than zero then it is positive and neutral if zero.